// @openzeppelin/relayer-sdk/dist/apis/plugins-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * OpenZeppelin Relayer API
 * OpenZeppelin Relayer API
 *
 * The version of the OpenAPI document: 1.2.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginsApi = exports.PluginsApiFactory = exports.PluginsApiFp = exports.PluginsApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * PluginsApi - axios parameter creator
 * @export
 */
const PluginsApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * Logs and traces are only returned when the plugin is configured with `emit_logs` / `emit_traces`. Plugin-provided errors are normalized into a consistent payload (`code`, `details`) and a derived message so downstream clients receive a stable shape regardless of how the handler threw.
         * @summary Execute a plugin and receive the sanitized result
         * @param {string} pluginId The unique identifier of the plugin
         * @param {PluginCallRequest} pluginCallRequest
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        callPlugin: (pluginId_1, pluginCallRequest_1, ...args_1) => __awaiter(this, [pluginId_1, pluginCallRequest_1, ...args_1], void 0, function* (pluginId, pluginCallRequest, options = {}) {
            // verify required parameter 'pluginId' is not null or undefined
            (0, common_1.assertParamExists)('callPlugin', 'pluginId', pluginId);
            // verify required parameter 'pluginCallRequest' is not null or undefined
            (0, common_1.assertParamExists)('callPlugin', 'pluginCallRequest', pluginCallRequest);
            const localVarPath = `/api/v1/plugins/{plugin_id}/call`
                .replace(`{${"plugin_id"}}`, encodeURIComponent(String(pluginId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearer_auth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(pluginCallRequest, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.PluginsApiAxiosParamCreator = PluginsApiAxiosParamCreator;
/**
 * PluginsApi - functional programming interface
 * @export
 */
const PluginsApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.PluginsApiAxiosParamCreator)(configuration);
    return {
        /**
         * Logs and traces are only returned when the plugin is configured with `emit_logs` / `emit_traces`. Plugin-provided errors are normalized into a consistent payload (`code`, `details`) and a derived message so downstream clients receive a stable shape regardless of how the handler threw.
         * @summary Execute a plugin and receive the sanitized result
         * @param {string} pluginId The unique identifier of the plugin
         * @param {PluginCallRequest} pluginCallRequest
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        callPlugin(pluginId, pluginCallRequest, options) {
            return __awaiter(this, void 0, void 0, function* () {
                var _a, _b, _c;
                const localVarAxiosArgs = yield localVarAxiosParamCreator.callPlugin(pluginId, pluginCallRequest, options);
                const localVarOperationServerIndex = (_a = configuration === null || configuration === void 0 ? void 0 : configuration.serverIndex) !== null && _a !== void 0 ? _a : 0;
                const localVarOperationServerBasePath = (_c = (_b = base_1.operationServerMap['PluginsApi.callPlugin']) === null || _b === void 0 ? void 0 : _b[localVarOperationServerIndex]) === null || _c === void 0 ? void 0 : _c.url;
                return (axios, basePath) => (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration)(axios, localVarOperationServerBasePath || basePath);
            });
        },
    };
};
exports.PluginsApiFp = PluginsApiFp;
/**
 * PluginsApi - factory interface
 * @export
 */
const PluginsApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.PluginsApiFp)(configuration);
    return {
        /**
         * Logs and traces are only returned when the plugin is configured with `emit_logs` / `emit_traces`. Plugin-provided errors are normalized into a consistent payload (`code`, `details`) and a derived message so downstream clients receive a stable shape regardless of how the handler threw.
         * @summary Execute a plugin and receive the sanitized result
         * @param {string} pluginId The unique identifier of the plugin
         * @param {PluginCallRequest} pluginCallRequest
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        callPlugin(pluginId, pluginCallRequest, options) {
            return localVarFp.callPlugin(pluginId, pluginCallRequest, options).then((request) => request(axios, basePath));
        },
    };
};
exports.PluginsApiFactory = PluginsApiFactory;
/**
 * PluginsApi - object-oriented interface
 * @export
 * @class PluginsApi
 * @extends {BaseAPI}
 */
class PluginsApi extends base_1.BaseAPI {
    /**
     * Logs and traces are only returned when the plugin is configured with `emit_logs` / `emit_traces`. Plugin-provided errors are normalized into a consistent payload (`code`, `details`) and a derived message so downstream clients receive a stable shape regardless of how the handler threw.
     * @summary Execute a plugin and receive the sanitized result
     * @param {string} pluginId The unique identifier of the plugin
     * @param {PluginCallRequest} pluginCallRequest
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof PluginsApi
     */
    callPlugin(pluginId, pluginCallRequest, options) {
        return (0, exports.PluginsApiFp)(this.configuration).callPlugin(pluginId, pluginCallRequest, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.PluginsApi = PluginsApi;
